### ISSUE TITLE
How to use fp16 training with masked operations

### ISSUE THREAD

**Issue by mfuntowicz** [NVIDIA/apex#93]
Hello ! 

I'm working on sequence training with CNNs and for this I've to operate some masked_fill operations over the padding, before softmax for example.

On float32 training, I'm masking with -1e20 value, and it seems to be training fine. Unfortunately, when training with float16 and masking with -1e15, amp loss scaling always returns "NaN" for gradients.

Do you have any idea how to combine masked_fill with amp ? 

Thanks, 
Morgan

**Comment by matthew-z** [comment]
```
>>> torch.Tensor([-1e15]).half()
tensor([-inf], dtype=torch.float16)
```

-1e15 is just too large for fp16, the smallest value in fp16 is -65504

**Comment by matthew-z** [comment]
If you are using AMP, it can automatically convert tensor to fp32 if is necessary. However, it seems that it is a bit tricky:


```python
>>> amp.init()
>>> torch.Tensor([-1]).cuda().half() + -1e20
tensor([-inf], device='cuda:0', dtype=torch.float16)       <-------AMP did nothing

>>> torch.Tensor([-1]).cuda().half() + torch.Tensor([-1e20]).cuda()
tensor([-1.0000e+20], device='cuda:0')                      <-------AMP converts the calculation into fp32
```

So, if you want amp to automatically fix this, just replace calculation like `x-1e20` with `x-torch.Tensor([1e20])`

**Comment by mcarilli** [comment]
@mfuntowicz I don't think Amp attempts to patch or do any conversions for masked_fill.  It's a straight copy op so amp doesn't see it as unsafe.  I think @matthew-z is correct in saying that it's simply a case of using a mask value that's not representable in FP16.  -1e15's magnitude is too large.  Try using -2^15 = -32768, that should work.  -65504 should also work.

@matthew-z your example is kind of subtle.  I think the reason the second example (`torch.Tensor([-1]).cuda().half() + torch.Tensor([-1e20]).cuda()`) works is that for operations with multiple inputs, Amp upcasts all inputs to match the widest incoming type.  In this case the type of the first operand to `+` is `half` and the type of the second operand is `float`, so Amp upcasts both inputs to `float` before performing the addition, and the patched `+` returns an output that's also `float`.

In the first example, the second input is a scalar, so Amp does not upcast the first input.  Also, it returns an output that's still `half`.

**Comment by matthew-z** [comment]
@mcarilli I get it. Thank you for the detailed explanation. 

I was surprised by the behavior since  PyTorch only raises error when float+half happens.

**Comment by mfuntowicz** [comment]
Thanks for all your answer. I made a mistake in my initial message as I'm not using -1e-15 for fp16 training but -(2 ** 15), which should get to the same fix @mcarilli is suggesting.

Anyway, even with -2 ^15 that gives me NaN gradients.
Below is the masked_softmax function I'm using : 

```python
def masked_softmax(x, mask, dim=-1):
    """
    ``torch.nn.functional.softmax(vector)`` does not work if some elements of ``vector`` should be
    masked.  This performs a softmax on just the non-masked portions of ``vector``.  Passing
    ``None`` in for the mask is also acceptable; you'll just get a regular softmax.
    We assume that both ``vector`` and ``mask`` (if given) have shape ``(batch_size, vector_dim)``.
    In the case that the input vector is completely masked, this function returns an array
    of ``0.0``. This behavior may cause ``NaN`` if this is used as the last layer of a model
    that uses categorical cross-entropy loss.
    """

    INF = (2 ** 15)

    if len(x.size()) == 3:
        mask = mask.unsqueeze(1)

    # Handle distributed settings
    if mask.size(-1) != x.size(-1):
        mask = pad(mask, (0, x.size(-1) - mask.size(-1), 0, 0), value=1)

    # To limit numerical errors from large vector elements outside the mask, we zero these out.
    x_ = x.masked_fill((1 - mask).byte(), -INF)
    return x_, softmax(x_, dim)

```

**Comment by mcarilli** [comment]
Hmm, that should be OK.  INF should be representable, and x_ should be upcast to float on entry to softmax.

I notice you're also returning `x_` directly.  Maybe the problem is coming from something else that uses `x_` later on.  For safety, you can try explicitly creating `x_` as float:
`x_ = x.float().masked_fill((1 - mask).byte(), -INF)`
This shouldn't hurt performance too much, because Amp will still convert `x_` back to half the next time it encounters a convolution or gemm.

**Comment by mfuntowicz** [comment]
I always use the softmax-ed values in the network, mainly to compute attentions giving no importance to padded entries.

The only place where ```x_``` is used directly is in the output of my model, bound to pytorch's ```cross_entropy``` loss.

Should I try to upscale as you mention the output of my model ?

**Comment by mcarilli** [comment]
With Amp, you shouldn't need to...but it can't hurt.  Did you actually try 
```
x_ = x.float().masked_fill((1 - mask).byte(), -INF)
```
?  I'm curious what the results might be.

**Comment by mfuntowicz** [comment]
Just tried with your suggestion but I still get the same : 

![image](https://user-images.githubusercontent.com/2241520/49899265-831d1380-fe5b-11e8-911c-49c3eead7628.png)
